Firmware/twi.c

```diff
@@ -0,0 +1,111 @@
+/*
+  twi.c - Stripped-down TWI/I2C library
+  Copyright (c) 2006 Nicholas Zambetti.  All right reserved.
+
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
+
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+
+  Modified 2012 by Todd Krein (todd@krein.org) to implement repeated starts
+*/
+
+#include <math.h>
+#include "Arduino.h" // for digitalWrite
+
+#ifndef cbi
+#define cbi(sfr, bit) (_SFR_BYTE(sfr) &= ~_BV(bit))
+#endif
+
+#ifndef sbi
+#define sbi(sfr, bit) (_SFR_BYTE(sfr) |= _BV(bit))
+#endif
+
+#include "twi.h"
+
+
+void twi_init(void)
+{
+  // activate internal pullups for twi.
+  digitalWrite(SDA, 1);
```

Comment:
Can we get rid of these Arduino functions and use fastio?

Reply:
Ironically I cannot find a "stock" pin definition for SCL/SDA which is directly usable with fastio.
I'd like to avoid defining one again, since I'm in the process of removing SWI2C_SCL/SDA...

The SDA definition in fastio.h itself is prefix with DIO already, which is not compatible with the macros. This is ... ironic? (I don't get the purpose of these definitions if they cannot be used with fastio!)

Arduino's own PIN_WIRE_SDA is braced with parenthesis, which again doesn't work with the macros.

Overall, since this is only done during init, and we're already using digitalWrite elsewhere, I wouldn't care.

Otherwise we need to define these pins ourselves in the board definition.

There's also the issue of cbi/sbi being duplicated, since I don't want to include the whole "Marlin.h" for these. Once we merge https://github.com/prusa3d/Prusa-Firmware/pull/2690 we can do some cleanup.

Reply:
As far as I understand the cbi and sbi macros are considered deprecated by avr-libc. Instead once should just do good old bitwise OR() and AND~().
Regardless of this, I think you can include something like <avr/io.h> which defines it as well.

Reply:
See https://www.eit.lth.se/fileadmin/eit/courses/edi021/avr-libc-user-manual/group__avr__sfr.html#:~:text=Porting%20programs%20that%20use%20the%20deprecated%20sbi%2Fcbi%20macros&text=The%20sbi%20and%20cbi%20macros,sfr%20%7C%3D%20_BV(bit)%20.

Reply:
Yeah, for one use you're right. I'll remove it.